streams/src/test/java/org/apache/kafka/streams/smoketest/ShutdownDeadlockTest.java

```diff
@@ -0,0 +1,100 @@
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.smoketest;
+
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.serialization.StringSerializer;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.kstream.ForeachAction;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.KStreamBuilder;
+
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+
+public class ShutdownDeadlockTest {
+
+    private final String kafka;
+    private final String zookeeper;
+
+    public ShutdownDeadlockTest(final String kafka,
+                                final String zookeeper) {
+
+        this.kafka = kafka;
+        this.zookeeper = zookeeper;
+    }
+
+    public void start() {
+        final String topic = "source";
+        final Properties props = new Properties();
+        props.setProperty(StreamsConfig.APPLICATION_ID_CONFIG, "shouldNotDeadlock");
+        props.setProperty(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        props.setProperty(StreamsConfig.ZOOKEEPER_CONNECT_CONFIG, zookeeper);
+        final KStreamBuilder builder = new KStreamBuilder();
+        final KStream<String, String> source = builder.stream(Serdes.String(), Serdes.String(), topic);
+
+        source.foreach(new ForeachAction<String, String>() {
+            @Override
+            public void apply(final String key, final String value) {
+                throw new RuntimeException("KABOOM!");
+            }
+        });
+        final KafkaStreams streams = new KafkaStreams(builder, props);
+        streams.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
+            @Override
+            public void uncaughtException(final Thread t, final Throwable e) {
+                System.exit(-1);
+            }
+        });
+
+        Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
+            @Override
+            public void run() {
+                streams.close(5, TimeUnit.SECONDS);
+            }
+        }));
+
+        final Properties producerProps = new Properties();
+        producerProps.put(ProducerConfig.CLIENT_ID_CONFIG, "SmokeTest");
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+
+        final KafkaProducer<String, String> producer = new KafkaProducer<>(producerProps);
+        producer.send(new ProducerRecord<>(topic, "a", "a"));
+        producer.flush();
+
+        streams.start();
+        while (true) {
```

Comment:
Would the try-wait-catch part not be sufficient? Why the loop (main thread will never receive a notify anyway? would it?) and why synchronized?